src/java.base/share/classes/sun/security/util/PBKDF2Parameters.java

```diff
@@ -0,0 +1,180 @@
+            ObjectIdentifier.of(KnownOIDs.HmacSHA1);
+
+    private String prfAlgo = "HmacSHA1";
+
+    public PBKDF2Parameters(DerValue keyDerivationFunc) throws IOException {
+
+        if (!pkcs5PBKDF2_OID.equals(keyDerivationFunc.data.getOID())) {
+            throw new IOException("PBKDF2 parameter parsing error: "
+                + "expecting the object identifier for PBKDF2");
+        }
+        if (keyDerivationFunc.tag != DerValue.tag_Sequence) {
+            throw new IOException("PBKDF2 parameter parsing error: "
+                + "not an ASN.1 SEQUENCE tag");
+        }
+        DerValue pBKDF2_params = keyDerivationFunc.data.getDerValue();
+        if (pBKDF2_params.tag != DerValue.tag_Sequence) {
+            throw new IOException("PBKDF2 parameter parsing error: "
+                + "not an ASN.1 SEQUENCE tag");
+        }
+        DerValue specified = pBKDF2_params.data.getDerValue();
+        // the 'specified' ASN.1 CHOICE for 'salt' is supported
+        if (specified.tag == DerValue.tag_OctetString) {
+            salt = specified.getOctetString();
+        } else {
+            // the 'otherSource' ASN.1 CHOICE for 'salt' is not supported
+            throw new IOException("PBKDF2 parameter parsing error: "
+                + "not an ASN.1 OCTET STRING tag");
+        }
+        iterationCount = pBKDF2_params.data.getInteger();
+
+        // keyLength INTEGER (1..MAX) OPTIONAL,
+        var ksDer = pBKDF2_params.data.getOptional(DerValue.tag_Integer);
+        if (ksDer.isPresent()) {
+            keyLength = ksDer.get().getInteger() * 8; // keyLength (in bits)
+        }
+
+        // prf AlgorithmIdentifier {{PBKDF2-PRFs}} DEFAULT algid-hmacWithSHA1
+        var prfDer = pBKDF2_params.data.getOptional(DerValue.tag_Sequence);
+        if (prfDer.isPresent()) {
+            DerValue prf = prfDer.get();
+            kdfAlgo_OID = prf.data.getOID();
+            KnownOIDs o = KnownOIDs.findMatch(kdfAlgo_OID.toString());
+            if (o == null || (!o.stdName().equals("HmacSHA1") &&
+                    !o.stdName().equals("HmacSHA224") &&
+                    !o.stdName().equals("HmacSHA256") &&
+                    !o.stdName().equals("HmacSHA384") &&
+                    !o.stdName().equals("HmacSHA512") &&
+                    !o.stdName().equals("HmacSHA512/224") &&
+                    !o.stdName().equals("HmacSHA512/256"))) {
+                throw new IOException("PBKDF2 parameter parsing error: "
+                        + "expecting the object identifier for a HmacSHA "
+                        + "pseudorandom function");
+            }
+            prfAlgo = o.stdName();
+            prf.data.getOptional(DerValue.tag_Null);
+            prf.data.atEnd();
+        }
+    }
+
+    /**
+     * Returns the salt.
+     *
+     * @return the salt. Returns a new array
+     * each time this method is called.
+     */
+    public byte[] getSalt() {
+        return this.salt.clone();
```

Comment:
I don't think you need to return a clone here as you are just using it internally and not mutating it or exposing it to applications that could modify it. I'll double-check as I review the rest of the code.

Reply:
fixed